[Android] Implement ScrollView sticky headers on Android

janicduplessis opened the pull request:
This adds support for sticky headers on Android. The implementation if based primarily on the iOS one (https://github.com/facebook/react-native/blob/master/React/Views/RCTScrollView.m#L272) and adds some stuff that was missing to be able to handle z-index, view clipping, view hierarchy optimization and touch handling properly.

Some notable changes:

Adds a `stickySectionHeaders` prop on `ListView` to control whether section headers are sticky. This is `true` on iOS and `false` on Android by default.

Add ChildDrawingOrderDelegate interface to allow changing the ViewGroup drawing order using ViewGroup#getChildDrawingOrder. This is used to change the content view drawing order to make sure headers are drawn over the other cells.

Add collapsableChildren prop that works like collapsable but applies to every child of the view. This is needed to be able to reference sticky headers by their indices otherwise some subviews can get optimized out and break indexes.

Make TouchTargetHelper take drawing order into consideration when finding the touch target.

Take translate into consideration when clipping subviews.

Fixes issues after the original PR (#9456) was reverted .

Test plan
Tested using the UIExplorer main menu on android.
Tested the Paging ListView example that uses LayoutAnimation to make sure it works.
Tested that touches get dispatched to the sticky header when over other touchable views.
Tested that touchables inside a sticky header work properly.

facebook-github-bot-0 commented:
By analyzing the blame information on this pull request, we identified @astreet and @sahrens to be potential reviewers.

janicduplessis commented:
This includes @astreet's fix for nodes that was also reverted as part of the original PR.

I think the main issue that caused the revert was the fact that I used to simply inverse the drawing order instead of changing the drawing order to only affect sticky headers. This adds quite a bit of complexity but it works well and should fix the bug that caused this to be reverted.

This also adds a fix for a bug that happened with LayoutAnimation where the sticked header would get animated in a weird way. To fix this I added a way for a view to opt out of LayoutAnimation.

Lastly this adds a prop on ListView to control whether sections headers should be sticky or not. Defaults to true on ios and false on android so it should make this PR not affect any `ListView` unless they opt in. This will still affect apps that use stickySectionIndices directly as it used to be a noop on android.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/layoutanimation/LayoutAnimationController.java:
```diff
@@ -66,6 +68,12 @@ public void reset() {
   }
 
   public boolean shouldAnimateLayout(View viewToAnimate) {
+    if (viewToAnimate instanceof ReactViewGroup) {
```
Not sure about this approach but layout animation didn't work properly at all here since the new position of the view isn't docked yet it gets animated in a weird way. What I did is simply add a flag to disable layout animation on ReactViewGroup.

astreet commented:
@facebook-github-bot import

astreet commented:
I'm gonna pass this to the adsmanager team so they can test it out :)

facebook-github-bot-4 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3879833) to review internal test results.

facebook-github-bot-4 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..9644a45b63f4dc94d5bb520ba54d80c56c92a0ad) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..9644a45b63f4dc94d5bb520ba54d80c56c92a0ad)

janicduplessis commented:
@astreet Oups it didn't build with buck, fixed.

facebook-github-bot-6 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/9644a45b63f4dc94d5bb520ba54d80c56c92a0ad..7eb571b4e991fb15e899ce24062dcff04163b0eb) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..7eb571b4e991fb15e899ce24062dcff04163b0eb)

nihgwu commented:
got time to test this PR, and it's really awesome, all the bugs I mentioned before are resolved
BTW, I think you can also add `stickySectionHeaders={true}` to `UIExplorerExampleList` to show power of this feature 😄

nihgwu commented:
unfortunately this new PR break my app completely 💔 , I'm making further investigating

nihgwu commented:
Every time I change the `stickySectionHeaders` I get this error(HMR is enabled)
![image](https://cloud.githubusercontent.com/assets/2595058/18608835/9e77f414-7d26-11e6-85c0-2ec0ddea77bc.png)
![image](https://cloud.githubusercontent.com/assets/2595058/18608849/173bf81e-7d27-11e6-9ba8-3d3b7d394799.png)

The header should be transparent but get white background
![image](https://cloud.githubusercontent.com/assets/2595058/18608867/72fe9526-7d27-11e6-96ec-3b46b5813e84.png)

The sticky header is overlapped by the content until another sticky header is triggered
![image](https://cloud.githubusercontent.com/assets/2595058/18608877/9ef0de96-7d27-11e6-9c0b-fe84bad67829.png)

janicduplessis commented:
@nigwu Could you send me the code that reproduces these issues?

nihgwu commented:
@janicduplessis The first issue should be introduced by recent commits to the master branch, the rest are only found in my app, but it works quite well with the old PR, quite strange, I've try a similar situation in UIExplorer, this new PR works well as expected. I'm struggling to find the cause

`ScollView sticky header on Android` and `Native Animated Event` are the last features I'd love to have, thanks for you great work

janicduplessis commented:
@nihgwu This PR changed the way we handle z-index by changing drawing order so it might be a bug with be new technique. First PR used to just reverse drawing order but that caused other issues, now it will keep the drawing order for non-sticky views and draw all sticky views at the end.

Lets say you have 5 rows:
1: sticky, 2: normal, 3: normal, 4: sticky, 5: normal

old PR:
5, 4, 3, 2, 1
new PR:
2, 3, 5, 1, 4

Code: https://github.com/facebook/react-native/pull/9957/files#diff-49a8f2a72edb3b8c02365cd11a8f2984R471

janicduplessis commented:
Thinking about this more I should not have to change the drawing order for all sticky headers, I could only change it for the one that is currenty sticked on top. It would probably simplify this quite a bit.

nihgwu commented:
I've notice that you change the drawing order, so I suspect that would be the cause for the different behave, I just don't know why the UIExplore is OK but my app isn't

nihgwu commented:
Just test with the newest PR, the first issue is another one and I've create an issue #9969 , the second one no longer exists, but the third is still there, should be something wrong with the drawing order logic

nihgwu commented:
@janicduplessis finally find a way to reproduce the third issue: set `NUM_SECTIONS = 1` in `ListViewPagingExample` , the second page is loaded when onEndReached in my app
I also encountered with force-close when trying to repeatedly scroll up and down quickly, no sure if has any relation with this PR

astreet commented:
Is this ready for review or no?

nihgwu commented:
I'm afraid not, there is still a issue need to be resolved, then would be fine

janicduplessis commented:
@astreet I'll fix the last issue @nihgwu found today then it will be ready fo review.

facebook-github-bot-6 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/368fc3b889ba11b92d5b0534a4615251f10f1604..63b25dff8cdd8fa454fbfd46bb4c8a3882a22291) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..63b25dff8cdd8fa454fbfd46bb4c8a3882a22291)

facebook-github-bot-7 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/63b25dff8cdd8fa454fbfd46bb4c8a3882a22291..cb99fae7d665817cc1e21ac788b8e422864a2418) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..cb99fae7d665817cc1e21ac788b8e422864a2418)

facebook-github-bot-1 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/cb99fae7d665817cc1e21ac788b8e422864a2418..1b2fef85444763a4fcb7e85e2f618784fd9eabb7) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..1b2fef85444763a4fcb7e85e2f618784fd9eabb7)

facebook-github-bot-8 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/368fc3b889ba11b92d5b0534a4615251f10f1604..63b25dff8cdd8fa454fbfd46bb4c8a3882a22291) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..63b25dff8cdd8fa454fbfd46bb4c8a3882a22291)

facebook-github-bot-3 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/63b25dff8cdd8fa454fbfd46bb4c8a3882a22291..1b2fef85444763a4fcb7e85e2f618784fd9eabb7) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..1b2fef85444763a4fcb7e85e2f618784fd9eabb7)

facebook-github-bot-3 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/1b2fef85444763a4fcb7e85e2f618784fd9eabb7..2940ccefd9d4220a27d76868e02784d03bfd3984) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..2940ccefd9d4220a27d76868e02784d03bfd3984)

facebook-github-bot-9 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/2940ccefd9d4220a27d76868e02784d03bfd3984..7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb)

janicduplessis commented:
@astreet Ok ,tested everything and it's working fine now so it's ready for review.

I ended up dropping collapseChildren as it just didn't work because the node's parent is always null when creating the view and I didn't really find another solution. It will be the user's responsability to add `collapsable={false}` if needed. This is not ideal but I ran out of ideas on how to make this work properly. I'll add a note in the doc about it.

@nihgwu If you have time to test this again it would be great. This PagingListView example with SECTIONS = 1 works fine now.

facebook-github-bot-4 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb..46832f48f0065789f730e0feccb5a39943ecdf7e) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..46832f48f0065789f730e0feccb5a39943ecdf7e)

facebook-github-bot-3 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/46832f48f0065789f730e0feccb5a39943ecdf7e..7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb)

facebook-github-bot-2 commented:
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9957/files/7443ad51fae58c95b2ed36ab9c2d8ea06d36a2cb..46832f48f0065789f730e0feccb5a39943ecdf7e) - [changes since last import](https://github.com/facebook/react-native/pull/9957/files/f5eb3e9fa9743594622701ca2ebdfacdbe43607c..46832f48f0065789f730e0feccb5a39943ecdf7e)

nihgwu commented:
@janicduplessis I'm afraid there is still something wrong, now the first row in the sections sticks instead of the header in my app, I will update will more detail later when I find the reason

janicduplessis commented:
@nihgwu Does it work if you add `collapsable={false}` on the top View you return from renderRow and renderHeader

nihgwu commented:
@janicduplessis add ` collapsable={false}` to renderHeader works for me, but don't you think that whatever the config is, only the section header should be sticky or not

janicduplessis commented:
@nihgwu Yea native events makes it so it doesn't send events to javascript. I didn't think about the ListView case where the onScroll event would still be needed by JS. We should probably make it so events are still sent to JS event if there is a native event then.

`collapsable={false}` is intended since I haven't found a solution that works properly with native view hierachy optimization on android so unless @astreet has some other alternatives I think we can keep it like this.

nihgwu commented:
OK, then this PR should be ready for review now, we can talk about the `ListView` stuff later when you send the PR to address the `Native Event` for `ListView`

astreet commented:
Having to manually add collapsable=false is just too easy to mess up -- I don't think we can afford to introduce this as an API in that case. Can you explain better why collapseChildren doesn't work? Was that responsible for some of the bugs we were seeing in the first version? Using a custom view wrapper instead of stickyHeaderIndices is looking like a better and better solution imo...

nihgwu commented:
IMO `collapsable=false` is not a new API, just to make sure the view hierarchy optimization on android won't break the `stickyHeaderIndices`, that's a cheap trick but can make it simple and be consistent with the iOS version.

janicduplessis commented:
@astreet The problem with `collapseChildren` is that we determine if a view is collapsable in `NativeViewHierarchyOptimizer#handleCreateView` which at this point the parent of the node that's being created is null because attaching the node to it's parent is done in `handleManageChildren` (those functions all end up being called from JS). `handleCreateView` needs to decide if it creates the native view or not but has no knowledge of it's parent view.

The only way I see that could work is by always creating the view in `handleCreateView` and to determine if it is collapsable in `handleManageChildren` and delete the native view we just created if it is the case but that feels like pretty wasteful and kind of throws away the perf optimisation for a very niche use case.

I guess the next step would be to try the custom view wrapper you suggested. My only concern with it is that I'd like to preserve the current API so I'll see if I can do something in JS to do that.

One last thing I can think of is to special case the `collapseChildren` prop in the react native renderer where it has full knowledge of the view hierarchy and adding `collapsable` prop to all childs of a view that has `collapsableChildren`.

astreet commented:
>The only way I see that could work is by always creating the view in handleCreateView and to determine if it is collapsable in handleManageChildren and delete the native view we just created if it is the case but that feels like pretty wasteful and kind of throws away the perf optimisation for a very niche use case.

Yeah, I'd really rather not do that.

The index API is just bad, at least on the native side: it's overloading concepts. There's a good chance you could keep the index API in JS and use it to auto-wrap child views in this special view wrapper. Another option might be to create a special ScrollView shadow node which uses onCollectExtraUpdates to send a list of view tags (which can be used to identify the native children of the scroll view: all the native views have their react tag returned from `getId()`). You would need to use onCollectExtraUpdates because you can only get the tags when you know all shadow node children have been added.

janicduplessis commented:
@astreet I like the idea of using shadow nodes to collect view tags, I'll try that.

astreet commented:
I'm trying to think, is there a chance that the view that's supposed to be sticky could be collapsed? If so, this won't work. We might need to automatically wrap the children at the sticky header indices in a View that has collapsable=false in JS to make sure this approach works.

janicduplessis commented:
Yea it should be possible that it happens. Something like
``` js
<View>
  <Text />
  <View style={{ backgroundColor: '#eee', height: 1 }} />
</View>
```
should get collapsed and is kind of common (pretty sure UIExplorer has something like this).

astreet commented:
Felix, can you keep an eye on this/help review while I'm away?

nihgwu commented:
any progress on this?

janicduplessis commented:
@nihgwu No, haven't had the time to look at this yet

astreet commented:
Hey, I was away during October. It looks like this is waiting on changes from you @janicduplessis?

janicduplessis commented:
Yes, I haven't had time to update this with what we discussed yet.

ouabing commented:
looking forward to this feature too :)

vitorebatista commented:
I need this!! :)

Ehesp commented:
Any updates @janicduplessis? Would love to see this in!

janicduplessis commented:
Planning to work on this in the next few weeks.

janicduplessis commented:
Closing this and will open a new PR as I've used a completely different approach.

janicduplessis commented:
Follow up: #11315
